### DISCUSSION TITLE
Algorithme und Datenstrukture

### DISCUSSION THREAD

**Discussion by ghost** [Estheryu991/SwissGerman_Dictionary#69] [Q&A]
Es gibt kei Grammatik. 

- [ ] Was isch ein [DP?]( https://cadmo.ethz.ch/education/lectures/HS22/DA/theory_exercises/Homework%2006.pdf) 

I möcht lieber ga Integer Ischriebe [lässig](https://www.huehner-hof.com/gefluegel/enten-halten/wo-kann-ich-enten-kaufen/)

![](https://www.huehner-hof.com/wp-content/uploads/2020/03/enten-kaufen.jpg)

> DP isch eine vielfältige und mächtige Klasse von Algorithmen. 
> Mit DP-Ansätzen kann man viele Probleme lösen, die uf den ersten Blick völlig unzugänglich erscheine.

> _DP ist eine vielfältige und mächtige Klasse von Algorithmen. Mit DP-Ansätzen kann man viele Probleme lösen, die auf den ersten Blick völlig unzugänglich erscheinen._

- [ ] Was isch ein Knapsack? 

- [ ] gattlech, gattlig: ordentlich, wohlanständig, schön oder freundlich anzuschauen 

Aufgabe 5.5 

Erraten eines Zahlenpaares (1 Punkt). Alice und Bob spielen das folgende Spiel:

- [ ] - Alice wählt zwei ganze Zahlen 1 ≤ a, b ≤ 1000, die sie geheim hält - Dann wiederholen Alice und Bob das Folgende:
- [ ] - Bob wählt zwei ganze Zahlen (a′, b′)
- [ ] - Ifa=a′ undb=b′,Bobgewinnt
- [ ] - Ifa>a′ undb>b′,AlicetellsBob'high!'
- [ ] - Ifa<a′ undb<b′,AlicetellsBob'low!'
- [ ] - Andernfalls gibt Alice Bob keinen Hinweis.

> Bob behauptet, er habe eine Strategie, um dieses Spiel in höchstens 12 Versuchen zu gewinnen. Beweisen Sie, dass es eine solche Strategie nicht geben kann.
> Tipp: Stellen Sie Bobs Strategie als Entscheidungsbaum dar. Jede Kante des Entscheidungsbaums entspricht einer von Alices Antworten, während jedes Blatt einem Gewinn für Bob entspricht.

- [ ] Tipp: Nachdem Sie den Entscheidungsbaum definiert haben, können Sie die Folge k0 = 1, kn+1 = 3kn + 1 betrachten und beweisen
- [ ] dass kn = 3n+1-1. Die Anzahl der Blätter im Entscheidungsbaum der Stufe n sollte mit kn zusammenhängen. 

- [ ] Was isch ein Matrix Chain? 

![](http://howimetmyoutfit.de/wp-content/uploads/2019/12/Laderach2-13-400x500.jpg)

  1 x 4 Schoggi 

- [ ] Was isch uere Letzte Multiplikation? 

![image](https://user-images.githubusercontent.com/117632901/200268563-6b43c800-dfb0-47fd-af37-3d44ef2c31ca.png)

Das isch optional. 

Jede Schoggi isch ei A_1 bis A_n, wir wollen uszurechne wie teuer es isch, 
dann nennen wir alles addieren. 

Horizontal isch auch Schoggi. 

Mir haan 3 Möglichkeiten. 

2 x 5 

![](http://mostlyaboutchocolate.com/wp-content/uploads/2020/11/laderach-world-choc-masters-winner-0.png)

### - [ ] Wie bauen mir das schneider? 

_Ich brauche ein Enke zmittag._ 

min {a_n, A_2.4, A_1.2 x A_1.2}

**Comment by ghost** [comment]
```
// EIn recursive Lösungen für subset summe
// problem
class GFG {

	// true zurück wenn gibt es ein subset
	// of set[] mit sum equal nach gegebene sum
	static boolean isSubsetSum(int set[],
							int n, int sum)
	{
		// Base Cases
		if (sum == 0)
			return true;
		if (n == 0)
			return false;

		// If last element is greater than
		// sum, then ignore it  Wenn das letzte Element größer ist als, Summe, dann wird es ignoriert
		if (set[n - 1] > sum)
			return isSubsetSum(set, n - 1, sum);

		/* else, check if sum can be obtained
		by any of the following
			(a) including the last element
			(b) excluding the last element */
		return isSubsetSum(set, n - 1, sum)
			|| isSubsetSum(set, n - 1, sum - set[n - 1]);
	}

	/* Driver code */
	public static void main(String args[])
	{
		int set[] = { 3, 34, 4, 12, 5, 2 };
		int sum = 9;
		int n = set.length;
		if (isSubsetSum(set, n, sum) == true)
			System.out.println("Subset gefunden"
							+ " mit gegebene Summe");
		else
			System.out.println("Kei Subset mit"
							+ " gegebene Summe");
	}
}



```

**Comment by esthicodes** [comment]
Wir wissen die Baum 

wir kuchen die Fibonazzi Zahl 
das wollen wir fertig mache 
mir wollen absetzen. 
Die Formel von F noch nie gse. 

F(n) = G(1+)

Die Fibonnazi Zahl isch so 
als mir Memorization gmacht haan.
Verwenden was mir genau berechnet haan. 

Wurzel 4 isch hdie n. 

Mir machen beide Seite und 

h<= 

das heisst wenn a von l halte, 
krieg genau was mir kaan. 

5 mal ankunft. 
entscheidunge mache. 

Und perfekt machen

Mit dem Einfügen die Einfügen und Technisch 
und ein Datenstrukturen haa. 

Vorher haan mir reingefügt + a_v bedingungen ige. 

Die neue Ufhgab gmacht haan. 
Und Mir haan alli repariert. 

zum biuspiu, alli ifüge haan, 
machen wir ifüge
ede Knote haan wir igbucht. 

Wenn mir andere gmacht haan, 
das wäre nie gmacht gha. 

Allgemeine Baum haan mir gmacht, irgendwo so was kei anig wie. 

Mir betracht die Igfüge, 
links oder rechts. 

Mir machen die Links ei, 
die Rechts 

Wenn mir öppis kaputt mache, 
welche Knoten muss man vorartzt mache. 
Wenn mir öppis gmacht haan. 

Was igfüge er woll das nöd. 
Welche Knoten möchte er mache. 

den Obesten. 
Weitere igfüge denn muss man öppis mache. 
Denn die vorgänge ghört, 
welche Teilbäume hesch du gmacht. 

Nur Teilbäume

Das hier isch bstimmte Fall, 
klick ich für reparierung. 
Wenn ich öppis ifolge, 
das isch nöd gändert. 

Teilbäum eins höher machen. 

Jetzt eins igfüge, (Run: links geh)
(Rechts machen mir links)
files seht so us. 
Vorher, haan mir knoten gmacht, 
links nichts

quasi nicht, igfügt worde, 
rechts wäre das igfügt. 

Die vorgänger, die Operation, 
das isch fertig gsi. 

Das isch die gschichte. 

über die Balance hier, kö

Vor links, fall unterscheidung. 

N - 1 : was isch den Fall? 

Bevor wäre minus 1. 

Muss mam zuerst öppis herstelle. 

zwei mal x, und denn haben wir igfügt. 

Höher von Teilbaum T isch ei Problem mit Back unter. 
Mal luege was tuet.
Die letzen Fall ischh hbal (p) 

Die Situation isch x igfügt die Situation 
jetzt die Teilbaum isch nöd ufgewachsen.

> Die Teilbaum ekann nicht ändern werde. 
> Die durch habe mir ufgewachsene Teilbaum gsi.

 Denn haan mir die Teilbaum, was haan mir gmacht?
Die höhe isch +1 ufgewachsen. 

Dann haan man ei Triangle 
 P het Vergangenheit, bed(p)≠ 0 

Dann mir mussen luege was Q isch. 

Die Frag isch die Fall unterscheidung je nach dem, die Balance von Q = +1 
Das heisst die Rechte vorher was machen wir so? 

Die Balance P isch repariert, und müssen wir weiter mache. 

Mega ufgewachsen. 

Die vorher isch minus eis. 
muss ich weiter obengehen. 

ABER DIE Teilbaume isch auch gewachsen. 
 
Dann haan man ei Unterschied. 
Denn isch ufgewachsen. 
Jetzt haan mir ei Problem gsi. 
Denn haan mir die 

Wil Q isch nid mehr aufgegangen.

P isch von Q gsi abo 
C hätte N vorgänger. 

Sind die nöd gli. 
Sie seht zeimlich gut us. 

bd(q) = 0 ;
bd(q) = 0;
fetch öppis. 

was soll man machen wenn RB nicht da bist? 
Red-Black-Tree

jetzt  hängt B isch unhängt erfüllt. 

die Ganzen haan mir gmacht
die Suchbaum haan mir gfüllt. 
Anderen File noch 
Plus einz. 
Diese füllt us 
Mir haan q und p 

hier hängt A. 

was möchtest du machen 

wie kann ich 
8.698 Millionen (2021) 
abonnenten gha uf Github? 

Die Teilbaum x isch igfügt. 

Die Referenzen isch Notation 

**Ich betrachte nur so quasi ufgewachsen bin.** 

`Jede Elementi gschreiben haan von O log(n)`
`a von log a und b, dann haan mir einfügen ist O(log n)`
ich werde zulese.
